lib/block_view.go

```diff
@@ -3603,6 +3621,87 @@ func (bav *UtxoView) _connectTransaction(txn *MsgDeSoTxn, txHash *BlockHash,
 	return utxoOpsForTxn, totalInput, totalOutput, fees, nil
 }
 
+// _connectFailingTransaction is used to process the fee and burn associated with the user submitting a failing transaction.
+// A failing transaction passes formatting validation, yet fails connecting to the UtxoView. This can happen for a number
+// of reasons, such as insufficient DESO balance, wrong public key, etc. With Revolution's Fee-Time block ordering, these
+// failing transactions are included in the blocks and their fees are burned. Initially blocks will include full transactions,
+// and in the future iterations blocks will only include the failing transaction hashes. This is done to prevent spamming
+// attacks on the network with seemingly valid transactions that never commit and just end up occupying the block space.
+// In addition, a major part of the effective fees of this transaction is burned with BMF. This makes spam attacks economically
+// disadvantageous. Attacker's funds are burned, to the benefit of everyone else on the network. BMF algorithm also computes
+// a utility fee, which is distributed to the block producer.
+func (bav *UtxoView) _connectFailingTransaction(txn *MsgDeSoTxn, blockHeight uint32, verifySignatures bool) (
+	_utxoOps []*UtxoOperation, _burnFee uint64, _utilityFee uint64, _err error) {
+
+	// Failing transactions are only allowed after ProofOfStake2ConsensusCutoverBlockHeight.
+	if blockHeight <= bav.Params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight {
+		return nil, 0, 0, fmt.Errorf("_connectFailingTransaction: Failing transactions " +
+			"not allowed before ProofOfStake2ConsensusCutoverBlockHeight")
+	}
+
+	// Sanity check the transaction to make sure it is properly formatted.
+	if err := CheckTransactionSanity(txn, blockHeight, bav.Params); err != nil {
+		return nil, 0, 0, errors.Wrapf(err, "_connectFailingTransaction: "+
+			"Problem checking txn sanity")
+	}
+
+	if err := ValidateDeSoTxnSanityBalanceModel(txn, uint64(blockHeight), bav.Params, bav.GlobalParamsEntry); err != nil {
+		return nil, 0, 0, errors.Wrapf(err, "_connectFailingTransaction: "+
+			"Problem checking txn sanity under balance model")
+	}
+
+	// Get the FailingTransactionBMFRateBasisPoints from the global params entry. We then compute the effective fee
+	// as: effectiveFee = txn.TxnFeeNanos * FailingTransactionBMFRateBasisPoints / 10000
+	gp := bav.GetCurrentGlobalParamsEntry()
+	failingTransactionRate := NewFloat().SetUint64(gp.FailingTransactionBMFRateBasisPoints)
+	failingTransactionRate.Quo(failingTransactionRate, NewFloat().SetUint64(10000))
+	effectiveFee, _ := NewFloat().Mul(failingTransactionRate, NewFloat().SetUint64(txn.TxnFeeNanos)).Uint64()
```

Comment:
Ideally you use integer math to compute this, which generally has fewer moving parts (no matissa/exponent to reason about for future configurations of floats). Feel free to create a re-usable helper function too

The uint64 math is just:
```
effectiveFee := txn.TxnFeeNanos * gp.FailingTransactionBMFRateBasisPoints / 10000
```

You may use uint256s if there's concern about overflow.

Reply:
Ah good suggestion, we should totally use integer math. I decided to use uint256, because uint64 could overflow in an undesirable manner. You can see this by running the code snippet below:

```
	x := uint64(1<<52 + 1000)
	y := uint64(1 << 12)
	fmt.Printf("x: (%v)\ny: (%v)\n(x*y)/10000: (%v)\n", x, y, (x*y)/uint64(10000))
```

uint64 overflow does modulo which can evaluate weirdly. In the above `x * 4096 / 10000` turns out to be `409`.